File: lib/features/boost/domain/services/purchase_verification_service.dart
import 'package:mevora/features/boost/domain/config/boost_product_config.dart';
import 'package:mevora/features/boost/domain/entities/purchase_record.dart';

enum VerificationOutcome {
  proceed,
  alreadyProcessed,
  invalidProduct,
  invalidUid,
  invalidTransaction,
  duplicateOtherUser,
}

class PurchaseVerificationDecision {
  const PurchaseVerificationDecision({
    required this.outcome,
    this.purchaseId,
  });

  final VerificationOutcome outcome;
  final String? purchaseId;

  bool get isAccepted =>
      outcome == VerificationOutcome.proceed ||
      outcome == VerificationOutcome.alreadyProcessed;
}

/// Validates product, uid, and transaction identity. Does **not** trust a
/// client "payment succeeded" flag. Store cryptographic verification happens
/// in Cloud Functions (ApplePurchaseVerifier / GooglePurchaseVerifier).
class PurchaseVerificationService {
  const PurchaseVerificationService({required this.config});

  final BoostProductConfig config;

  static String purchaseIdFor({
    required PurchasePlatform platform,
    required String transactionId,
  }) {
    return '${platform.name}_$transactionId';
  }

  PurchaseVerificationDecision decide({
    required String uid,
    required String productId,
    required String transactionId,
    required PurchasePlatform platform,
    PurchaseRecord? existing,
  }) {
    if (uid.trim().isEmpty) {
      return const PurchaseVerificationDecision(
        outcome: VerificationOutcome.invalidUid,
      );
    }
    if (transactionId.trim().isEmpty) {
      return const PurchaseVerificationDecision(
        outcome: VerificationOutcome.invalidTransaction,
      );
    }
    if (!config.isAllowedProductId(productId, platform)) {
      return const PurchaseVerificationDecision(
        outcome: VerificationOutcome.invalidProduct,
      );
    }

    final purchaseId = purchaseIdFor(
      platform: platform,
      transactionId: transactionId,
    );

    if (existing != null) {
      if (existing.userId != uid) {
        return PurchaseVerificationDecision(
          outcome: VerificationOutcome.duplicateOtherUser,
          purchaseId: purchaseId,
        );
      }
      if (existing.status == PurchaseRecordStatus.verified) {
        return PurchaseVerificationDecision(
          outcome: VerificationOutcome.alreadyProcessed,
          purchaseId: purchaseId,
        );
      }
    }

    return PurchaseVerificationDecision(
      outcome: VerificationOutcome.proceed,
      purchaseId: purchaseId,
    );
  }
}

File: functions/src/boost/purchaseVerificationService.ts
import {purchaseDocId} from "./config.js";
import type {BoostPack} from "./catalog.js";
import type {ApplePurchaseVerifier} from "./applePurchaseVerifier.js";
import type {GooglePurchaseVerifier} from "./googlePurchaseVerifier.js";
import type {
  PurchaseLedger,
  StorePlatform,
  StoreVerificationResult,
  VerifyBoostRequest,
} from "./types.js";

export type VerificationDecision =
  | {outcome: "proceed"; purchaseId: string; store: StoreVerificationResult}
  | {outcome: "alreadyProcessed"; purchaseId: string}
  | {outcome: "invalidProduct"}
  | {outcome: "invalidUid"}
  | {outcome: "invalidTransaction"}
  | {outcome: "duplicateOtherUser"; purchaseId: string}
  | {outcome: "storeInvalid"}
  | {outcome: "storeUnavailable"};

export class PurchaseVerificationService {
  constructor(
    private readonly apple: ApplePurchaseVerifier,
    private readonly google: GooglePurchaseVerifier,
  ) {}

  async verify(params: {
    uid: string;
    request: VerifyBoostRequest;
    pack?: BoostPack | null;
    existing?: PurchaseLedger | null;
  }): Promise<VerificationDecision> {
    const {uid, request, existing, pack} = params;
    if (!uid.trim()) {
      return {outcome: "invalidUid"};
    }
    if (!request.transactionId?.trim()) {
      return {outcome: "invalidTransaction"};
    }
    const platform: StorePlatform = request.platform === "ios" ? "ios" : "android";
    if (!pack || pack.productId !== request.productId || (pack.durationMs < 1 && pack.boostCount < 1)) {
      return {outcome: "invalidProduct"};
    }

    const purchaseId = purchaseDocId(platform, request.transactionId);
    if (existing) {
      if (existing.userId !== uid) {
        return {outcome: "duplicateOtherUser", purchaseId};
      }
      if (existing.status === "verified") {
        return {outcome: "alreadyProcessed", purchaseId};
      }
    }

    const store =
      platform === "ios" ? await this.apple.verify(request) : await this.google.verify(request);
    if (!store.ok) {
      return {outcome: store.error === "unavailable" ? "storeUnavailable" : "storeInvalid"};
    }
    if (store.productId !== request.productId) {
      return {outcome: "invalidProduct"};
    }
    return {outcome: "proceed", purchaseId, store};
  }
}
